feat(post): #10 Post 도메인 CRUD 및 테스트 코드 추가

sumssumm opened the pull request:
### 📌 PR 타입(하나 이상의 PR 타입을 선택해주세요)
- [x] 기능 추가
- [ ] 기능 삭제
- [ ] 버그 수정
- [x] 의존성, 환경 변수, 빌드 관련 코드 업데이트

### 📌 관련 이슈
- [KeepGoing-web/Keepgoing-BE/issues/10]

### ✨ 반영 브랜치
feat/10 -> develop

### 📝 변경 사항
- [x] Post 도메인에 User(author) 연관관계 적용 및 기본 CRUD 기능 구현
- [x] Post DTO 개선 (Create/Update/Response 정리 및 검증 애노테이션 보강)
- [x] PostService / PostController 로직 정리 및 예외 상황 처리 연동
- [x] GlobalExceptionHandler와 연계되는 권한 실패(POST_ACCESS_DENIED) 및 게시글 없음(POST_NOT_FOUND) 케이스 반영
- [x] Post 도메인/서비스/컨트롤러 단위 테스트 및 통합 테스트 추가  
  - PostTest: 도메인 메서드(create, update, softDelete, 권한 검증 등)
  - PostServiceTest: 정상/예외 플로우 검증
  - PostControllerTest: MockMvc 기반 HTTP 레벨 성공/실패 응답 검증
  - PostRepositoryTest: 작성자/공개범위 조회 및 soft delete 동작 확인
- [x] data.sql을 이용한 테스트/로컬 환경 초기 데이터 설정 (기본 유저 및 게시글)
- [x] JPA Auditing 활성화 및 created_at / updated_at 자동 관리 설정

### ➕ 추가 메모
- 현재 인증/인가 로직이 적용되지 않은 상태라, PostController에서는 `authorId = 1L` 을 하드코딩하여 사용하고 있습니다.  
  추후 Spring Security 및 인증 기능 적용 시, 로그인 유저 정보를 통해 authorId를 주입하도록 변경할 예정입니다.
- Post 조회/삭제 시 soft delete 전략을 사용하고 있으며, `@Where` 를 통해 deleted_at이 설정된 데이터는 기본 조회에서 제외되도록 구성했습니다.
- Post 기능 1차 구현 및 테스트를 마친 상태로, 이후 PR에서 페이징 및 추가 기능(공개 글 목록 API 등)을 확장할 계획입니다.

### 🐛 테스트 결과 (테스트 결과가 있다면 넣어주세요)
- [x] `./gradlew test` 전체 테스트 통과
- [x] IntelliJ HTTP Client를 활용한 Post API 수동 검증  
  - POST /api/v1/posts: 게시글 생성 확인  
  - GET /api/v1/posts/{id}, /me: 단일/목록 조회 동작 확인  
  - PUT /api/v1/posts/{id}: 수정 후 응답 내용 검증  
  - DELETE /api/v1/posts/{id}: soft delete 적용 및 이후 조회 시 예외 응답(POST_NOT_FOUND) 확인  
  - 권한 실패 시 403 + `POST_ACCESS_DENIED` 응답 구조 확인

sumssumm committed:
feat: #10 게시글 CRUD 및 공통 응답/예외 처리 연동

- Post 엔티티/DTO/Repository/Service/Controller 추가
- ApiResponse<T> 적용으로 Post API 공통 응답 포맷 통일
- ErrorCode에 POST_NOT_FOUND, POST_ACCESS_DENIED 추가
- GlobalExceptionHandlerTest 일부 수정
- Post HTTP Client 스크립트 및 단위 테스트 추가

sumssumm committed:
refactor(post): #10 User 엔티티와 연관관계 매핑

- Post 엔티티: authorId 필드를 User 엔티티와 ManyToOne 연관관계로 변경
- PostService: UserRepository를 사용하여 작성자(User)를 조회하도록 수정
- PostRepository: 변경된 엔티티 관계에 맞춰 findByAuthor_Id... 쿼리 메서드 사용
- ErrorCode: Post 관련 에러 코드(POST_NOT_FOUND, POST_ACCESS_DENIED) 추가

sumssumm committed:
chore: #10 JPA Auditing 활성화 및 데이터베이스 초기화 설정

- @EnableJpaAuditing 어노테이션 추가로 JPA Auditing 기능 활성화
- application.yml: 데이터베이스 스키마 생성 및 초기화 관련 JPA 설정 추가

sumssumm committed:
refactor: #10 Post DTO 개선 및 기타 수정

- Post DTO: Builder 패턴 적용하여 객체 생성 유연성 및 가독성 향상
- ERD: Post 관련 테이블 및 컬럼명 변경사항 반영
- Test: GlobalExceptionHandlerTest의 테스트 설정을 @SpringBootTest로 변경

sumssumm committed:
test: #10 Post 기능 관련 테스트 코드 구현 및 개선

- PostControllerTest: 컨트롤러 계층 테스트 코드 작성
- PostRepositoryTest: 레포지토리 계층 테스트 코드 작성
- PostServiceTest: updatePost 성공 케이스의 검증 로직 상세화
- PostDomainTest: validateAuthor 실패 케이스의 검증 로직 상세화

sumssumm committed:
test: #10 테스트 데이터 및 권한 실패 케이스 추가

- data.sql: 권한 테스트를 위한 '다른 유저' 및 '다른 유저 게시글' 데이터 추가
- post-api.http: HTTP Client 스크립트에서 변수를 사용하도록 개선하고, 권한 실패 테스트 케이스 추가
- PostControllerTest: 게시글 수정/삭제 시, 작성자가 아닌 경우 403 에러를 반환하는지 검증하는 테스트 케이스 추가

sumssumm committed:
test: #10 Repository 테스트 보강 및 인증 TODO 추가

- PostController: 하드코딩된 authorId에 인증 기능 적용을 위한 TODO 주석 추가
- PostRepositoryTest: findByVisibility 및 soft delete 관련 테스트 케이스 추가

ClszK commented on src/main/java/com/keepgoing/keepgoing/post/controller/PostController.java:
```diff
@@ -0,0 +1,66 @@
+import lombok.RequiredArgsConstructor;
+import org.springframework.web.bind.annotation.*;
+
+import java.util.List;
+
+@RestController
+@RequestMapping("/api/v1/posts")
+@RequiredArgsConstructor
+public class PostController {
+
+    private final PostService postService;
+
+    /**
+     * 포스트 생성
+     */
+    @PostMapping
+    public ApiResponse<PostResponse> createPost(@RequestBody @Valid PostCreateRequest request) {
+        Long authorId = 1L; // TODO: Security 붙으면 현재 로그인 유저로 교체
+        return ApiResponse.success(postService.createPost(authorId, request));
+    }
```
현재 return 값의 HttpStatus 응답코드가 200으로 반환되는데 201(새 리소스가 생성됨)을 반환하도록
```java
    @PostMapping
    public ResponseEntity<ApiResponse<PostResponse>> createPost2(@RequestBody @Valid PostCreateRequest request) {
        Long authorId = 1L; // TODO: Security 붙으면 현재 로그인 유저로 교체
        return ResponseEntity.status(HttpStatus.CREATED)
                .body(ApiResponse.success(postService.createPost(authorId, request)));
    }
```

이렇게 변경하는 게 어떨까요?

ClszK commented on src/main/java/com/keepgoing/keepgoing/post/domain/Post.java:
```diff
@@ -0,0 +1,139 @@
+@Entity
+@Table(name = "posts")
+@EntityListeners(AuditingEntityListener.class)
+@Where(clause = "deleted_at IS NULL") // soft delete 기본 필터
```
🔴 **HIGH: @Where → @SQLRestriction 마이그레이션 권장**

`@Where`는 Hibernate 6.3부터 deprecated 되었습니다 (Spring Boot 3.2+).

**권장 수정:**
```java
import org.hibernate.annotations.SQLRestriction;

@SQLRestriction("deleted_at IS NULL")
```

**근거:** 향후 Hibernate 버전 업그레이드 시 경고/오류 발생 가능

ClszK commented on src/main/java/com/keepgoing/keepgoing/post/controller/PostController.java:
```diff
@@ -0,0 +1,66 @@
+     * 포스트 생성
+     */
+    @PostMapping
+    public ApiResponse<PostResponse> createPost(@RequestBody @Valid PostCreateRequest request) {
```
🔴 **HIGH: POST 생성 시 201 Created 반환 권장**

REST API 관례상 리소스 생성은 **201 Created** 반환이 표준입니다.
Auth API(signup)에서도 201 반환 중이므로 일관성을 위해 수정 권장합니다.

**권장 수정:**
```java
@PostMapping
public ResponseEntity<ApiResponse<PostResponse>> createPost(...) {
    return ResponseEntity.status(HttpStatus.CREATED)
            .body(ApiResponse.success(...));
}
```

ClszK commented on src/test/java/com/keepgoing/keepgoing/post/controller/PostControllerTest.java:
```diff
@@ -0,0 +1,301 @@
+    @Autowired
+    ObjectMapper objectMapper;
+
+    @MockBean
```
🟡 **MEDIUM: @MockBean → @MockitoBean 권장**

Spring Boot 3.4+에서는 `@MockitoBean` 사용이 권장됩니다.
Auth 도메인 테스트에서는 이미 `@MockitoBean` 사용 중이므로 프로젝트 내 일관성을 위해 변경 권장합니다.

```java
import org.springframework.test.context.bean.override.mockito.MockitoBean;

@MockitoBean
PostService postService;
```

ClszK commented on src/main/java/com/keepgoing/keepgoing/post/dto/PostResponse.java:
```diff
@@ -0,0 +1,39 @@
+
+    private Long id;
+    private Long authorId;
+//    private String authorName;
```
🟡 **MEDIUM: 주석 처리된 코드 제거 권장**

주석 처리된 코드는 혼란을 유발합니다. 삭제하거나, 필요하다면 TODO 주석으로 명시해주세요.

ClszK commented on src/main/java/com/keepgoing/keepgoing/post/controller/dto/PostUpdateRequest.java:
```diff
@@ -0,0 +1,26 @@
+    @NotBlank
+    private String content;
+
+    @NotNull
```
🟡 **MEDIUM: Create/Update Request visibility 일관성 확인**

- PostCreateRequest: @Builder.Default (기본값 PRIVATE, null 허용)
- PostUpdateRequest: @NotNull (null 불허)

의도된 차이인지 확인 부탁드립니다!

sumssumm replied:
Post 생성 시에는 visibility를 선택적으로 받고, 값을 보내지 않으면 기본값 PRIVATE으로 저장하려는 의도가 있어서
PostCreateRequest 에는 @NotNull을 두지 않았습니다.

반대로, Post 수정 시에는 전체 필드를 교체하는 PUT 스타일로 설계해서
title / content / visibility / aiCollectable 을 항상 명시적으로 보내도록 하기 위해
PostUpdateRequest.visibility 에는 @NotNull을 사용했습니다.

다만, 말씀해주신 것처럼 Create/Update 간 일관성이 떨어져 보일 수는 있어서
- 생성에서도 visibility를 필수로 받도록 통일하거나,
- 아니면 둘 다 optional로 두고 PATCH 스타일로 바꾸는 것도 검토해보겠습니다.

ClszK reviewed (commented):
코드 리뷰 완료했습니다. 인라인 코멘트들을 확인해주세요. 전체적으로 잘 작성된 PR입니다! 🎉

sumssumm committed:
refactor: #10 Post 웹 DTO 패키지 및 정의 정리

- post/dto 패키지를 controller/dto 하위로 이동
- PostCreateRequest/PostUpdateRequest/PostResponse 패키지 변경
- 컨트롤러 및 테스트에서 변경된 DTO 패키지 경로로 import 수정

sumssumm committed:
refactor: #10 Post 도메인 및 서비스 로직 정리

- PostService가 웹 DTO 대신 도메인 값/엔티티를 사용하도록 변경
- PostService가 Post 엔티티를 반환하도록 시그니처 정리
- PostServiceTest/PostTest를 새로운 구조에 맞게 수정
- 에러 코드 정의 보완

sumssumm committed:
refactor: #10 PostController 응답 형식 및 상태 코드 정리

- PostController 전 메서드에 ResponseEntity<ApiResponse<...>> 적용
- 생성 시 201 Created, 조회/수정 시 200 OK, 삭제 시 204 No Content 반환
- PostControllerTest에서 변경된 응답 형식에 맞게 검증 로직 수정

sumssumm committed:
chore: 로컬 개발용 JPA 및 SQL 초기화 설정 조정

sumssumm committed:
refactor: #10 PostRepository에 @EntityGraph 적용 및 테스트 보강
